### PR TITLE
[megatron] fix: resolve backward propagation error in megatron_actor due to shared logits tensor in-place modification

### PR DESCRIPTION
**Pull request by HelloWorld686**
### What does this PR do?
Fixes gradient computation conflict in `verl/workers/actor/megatron_actor.py` when entropy regularization is enabled:
- **Root Cause**: The entropy calculation `entropy = vocab_parallel_entropy(logits)` fails during backward propagation because `log_probs = vocab_parallel_log_probs_from_logits(logits, label)` performs in-place modifications on the logits tensor earlier in the code. This corrupts the original computation graph needed for gradient calculation.
- **Fix**: Decouples tensor dependencies by cloning logits before entropy calculation to preserve the original computation graph while maintaining existing log_probs computation flow.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

1. Run modified training script:
```bash
examples/ppo_trainer/run_qwen2-7b_math_gsm8k_megatron.sh \
--actor_rollout_ref.actor.entropy_coeff=0.01
```

2. The following error is observed (before repair):
image

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [x] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ).

### CODE REVIEW

File: verl/workers/actor/megatron_actor.py [hunk 1]
```diff
@@ -522,10 +522,11 @@ def logits_processor(logits, label, label_mask):
                     assert logits.shape[:2] == label.shape[:2]
                     assert label.shape == label_mask.shape
                     ret = {}
+                    logits_bak = logits.clone()
                     if calculate_entropy:
                         entropy = vocab_parallel_entropy(logits)
                         ret["entropy"] = entropy
-                    log_probs = vocab_parallel_log_probs_from_logits(logits, label)
+                    log_probs = vocab_parallel_log_probs_from_logits(logits_bak, label)
                     log_probs = log_probs.masked_fill(~label_mask, 0.0)
```
[the remark this answers is not in the document]
Comment (ETOgaosion) [reply verl/workers/actor/megatron_actor.py]: Thanks for contribution, @HelloWorld686 Could you consider this advice?

**Comment by ETOgaosion** [comment]
Notice that here 

https://github.com/volcengine/verl/blob/d9a6a31c8d92030f456816d7e6b0c56b117fb192/verl/utils/megatron/tensor_parallel.py#L133

And

https://github.com/volcengine/verl/blob/d9a6a31c8d92030f456816d7e6b0c56b117fb192/verl/utils/megatron/tensor_parallel.py#L136-L137

`vocab_parallel_logits` shall remain the same, so current logic seems OK?

**Comment by HelloWorld686** [comment]
> Notice that here
> 
> https://github.com/volcengine/verl/blob/d9a6a31c8d92030f456816d7e6b0c56b117fb192/verl/utils/megatron/tensor_parallel.py#L133
> 
> And
> 
> https://github.com/volcengine/verl/blob/d9a6a31c8d92030f456816d7e6b0c56b117fb192/verl/utils/megatron/tensor_parallel.py#L136-L137
> 
> `vocab_parallel_logits` shall remain the same, so current logic seems OK?

During the forward pass of `vocab_parallel_entropy`, it does not alter the values of the logits; however, it saves the logits for use in backpropagation. In contrast, `vocab_parallel_log_probs_from_logits` modifies the logits in-place, which can cause errors during the backpropagation step of `vocab_parallel_entropy`.

File: verl/workers/actor/megatron_actor.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (HelloWorld686) [reply verl/workers/actor/megatron_actor.py]: Good idea, thanks

**Comment by vermouth1992** [comment]
In megatron, entropy is only computed for observation, not for regularization. logits.clone consumes too much memory and is not affordable. We need a kernel to make entropy regularization happen

**Comment by ETOgaosion** [comment]
#2210 has been merged as mcore side fused kernel function.

Shall we fix the correctness here?

**Comment by HelloWorld686** [comment]
> #2210 has been merged as mcore side fused kernel function.
> 
> Shall we fix the correctness here?

Thank you for your feedback. While memory efficiency is important, ensuring functional correctness should be our top priority—especially since the use_fused_kernels path doesn’t cover all use cases. The clone operation prevents bugs in the non-use_fused_kernels path. (We can’t just force-enable fused kernels to avoid bugs, can we?)

What do you think about prioritizing correctness across all scenarios first, then optimizing memory afterward?

File: verl/workers/actor/megatron_actor.py [hunk 2]
```diff
@@ -523,9 +523,12 @@ def logits_processor(logits, label, label_mask):
                     assert label.shape == label_mask.shape
                     ret = {}
                     if calculate_entropy:
+                        logits_bak = logits.clone()
```
Comment (ETOgaosion) [inline verl/workers/actor/megatron_actor.py:533]: @HelloWorld686 Could you add a `warnings.warn` here to let users know it is better to enable fused kernel to allow memory efficient calculation.

Comment (HelloWorld686) [reply verl/workers/actor/megatron_actor.py:533]: OK, modified

**Comment by HelloWorld686** [comment]
@ETOgaosion , I've rebased my PR with the latest code and now some of the tests have passed, including the one that previously failed. However, the PR run was cancelled a few days ago and there hasn't been any further activity. Could you please re-review the PR when you have a chance? Thank you!
